### PR TITLE
[MHLO-Linalg] Lower mhlo.convolution to linalg when padding is non zero.

### PR DESCRIPTION
**Pull request by arnab-polymage**
Lower mhlo.convolution to linalg when padding is non zero.

### CODE REVIEW

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc [hunk 1]
```diff
@@ -1729,6 +1761,27 @@ struct NormalConvOpConversion : public OpConversionPattern<mhlo::ConvOp> {
...
+      pad.resize(pad.size() + 2, 0);
+      // Pad the given input.
+      input = applyPad(loc, input, pad, zero_attr, rewriter);
+      // Create conv op using the padded input.
+      auto convOp = rewriter.create<linalg::Conv2DNhwcHwcfOp>(
+          loc, result_type, ValueRange{input, filter}, ValueRange{zero_tensor},
+          strides, dilations, PruneAttributeList(op));
+      rewriter.replaceOp(op, convOp.getOperation()->getResults());
+      return success();
+    }
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc:1783]: Why the special case for rank 4 only?

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc [hunk 2]
```diff
@@ -1729,6 +1761,27 @@ struct NormalConvOpConversion : public OpConversionPattern<mhlo::ConvOp> {
     // TODO(ataei): Only support dilated kernel right now. We need to consider
     // input dilation for deconvolution cases.
     Attribute dilations = op.rhs_dilationAttr();
+
+    // Check if padding is zero.
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc:1765]: I think you should improve the comment here, it should be a few lines to describe what's happening here, possibly explaining how the padding is handled on the input.

**Comment by joker-eph** [comment]
If I read the patch correctly: we are currently lowering these test cases but incorrectly so? 
So this is a bug fix?

Can you elaborate in the pull-request description?

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc [hunk 3]
```diff
@@ -1679,6 +1679,44 @@ struct PadOpConversion : public OpConversionPattern<mhlo::PadOp> {
   }
 };
 
+/// Extract element values from `attr` and store them in `arrayValues`.
+static void getValuesFromIntAttribute(DenseIntElementsAttr attr,
+                                      SmallVector<int64_t, 4>& arrayValues) {
+  for (auto val : attr.getValues<APInt>())
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc:1685]: You should be able to do something like:

```suggestion
  for (auto val : attr.getValues<int64_t>())
```

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc (hunk 1 above)
Comment (arnab-oss) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc:1776]: Sure. We can extend this for rank 3 and rank 4 too

**Review by joker-eph — COMMENTED** [review]
Sorry: I wrote some minor inline comment last time and didn't send them out...

LG overall.

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc [hunk 4]
```diff
@@ -1679,6 +1679,44 @@ struct PadOpConversion : public OpConversionPattern<mhlo::PadOp> {
...
+    arrayValues.push_back(val.getSExtValue());
+}
+
+/// Apply padding values stored in `pad` to `input`.
+static Value applyPad(Location loc, Value input, ArrayRef<int64_t> pad,
+                      Attribute padAttr, OpBuilder& rewriter) {
+  // Input should be padded if necessary.
+  if (llvm::all_of(pad, [](int64_t p) { return p == 0; })) return input;
```
[the remark this answers is not in the document]
Comment (arnab-oss) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc:1692]: Actually this condition is redundant....I'm already checking this condition in the caller function

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc (hunk 3 above)
Comment (arnab-oss) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc]: I think this is a useful helper function, which may have future uses. It no way depends on the lowering which I'm implementing.

**Review by joker-eph — APPROVED** [review]

**Pull request merged**